### ISSUE TITLE
Debug ACT for Native Triggers

### ISSUE THREAD

**Issue by jjscheel** [riscv-admin/dev-partners#7]
### Technical Group

Debug TG

### ratification-pkg

Debug

### Technical Liaison

Tim Newsom

### Task Category

Arch Tests

### Task Sub Category

- [ ] gcc
- [ ] binutils
- [ ] gdb
- [ ] intrinsics
- [ ] Java
- [ ] KVM
- [ ] ld
- [ ] llvm
- [ ] Linux kernel
- [ ] QEMU
- [ ] Spike

### Ratification Target

3Q2023

### Statement of Work (SOW)

[SOW link](https://docs.google.com/document/d/1ZeHeSBnCTknoJMt1dRrH66Bg7aI74oBMkxamsmXaPTs/)

**SOW Signoffs:** (delete those not needed)
  - Task group liaison sign-off date:
  - Development partner sign-off date:
  - ACT SIG sign-off date (if ACT work):

### Waiver

- [ ] Freeze
- [ ] Ratification

**Comment by jjscheel** [comment]
Per 4/11 email from Prasanna:
> Debug : sangee status at where we were last, nothing progressed from our last meeting

**Comment by jjscheel** [comment]
Will keep this on the Agenda for discussion as we work to find a way to offload this effort based on Zfinx, Zfh running long and new work in Zdinx and Zhinx.

I've started a discussion with @timsifive and @pdonahue-ventana to explore contingencies.

**Comment by jjscheel** [comment]
We discussed this today in the Debug TG call without any takers.  So, we shall see.  Tim and Paul understand that without a volunteer, this may take us some time to get to.

**Comment by jjscheel** [comment]
@timsifive, @pdonahue-ventana, any volunteers to help here yet?

**Comment by timsifive** [comment]
I haven't heard from anyone.

**Comment by jjscheel** [comment]
Comment from most recent ARC minutes (July 10) - [link](https://lists.riscv.org/g/tech-chairs/message/1507)

>-  **Debug 1.0:** Review has started of the latest submitted spec, but just review of all the changes since the last Debug 1.0 draft spec was submitted, reviewed, and approved.   Review of this has started, with initial discussion of one of the more notable backward-incompatible changes (between v1.0 and the ratified v0.13).     This relates with abstract commands and the data registers to support more efficient FPGA implementations.  After further discussion in next week's ARC meeting, John will be providing ARC feedback and guidance to the Debug TG.  Likely other items of feedback that arise will also be provided.
>- Review of the latest Debug and Pointer Masking specs (along with the S*deleg extensions) are expected to continue next week.

Also note, this remains on the list of future DevPartners activities.  It will be addressed AFTER Pointer Masking and the Floating Point gap work work (Zdinx, Zhinx).

**Comment by jjscheel** [comment]
Assigning back to IITM now that Zfinx/Zfh/Zdinx/Zhinx are wrapping up.

@ptprasanna, please dust off the coverage matrix and begin working on tests using Spike or QEMU.  Technical questions can be directed to @timsifive.

**Comment by jjscheel** [comment]
@ptprasanna, this item is getting close to Freeze.  As such, I'd like to understand the progress we've made in our next meeting. Please either attend or post status at your earliest convenience.

**Comment by anuani21** [comment]
Debugger existing test logs are placed in the below links:

https://gitlab.com/ptprasanna/actreports/-/tree/main/debugger/spike%2032?ref_type=heads in spike32.
https://gitlab.com/ptprasanna/actreports/-/tree/main/debugger/spike%2064?ref_type=heads in spike64.

**Comment by jjscheel** [comment]
This is good news, @ptprasanna and @anuani21!

@timsifive, note the progress.

**Comment by timsifive** [comment]
Note that all tests in riscv-tests/debug test debugging through an external debugger. None of them test native trigger behavior, where a trigger is set and when it is hit the hart takes a debug exception.

**Comment by jjscheel** [comment]
@ptprasanna, I would like an update in the issue, by the next meeting on December 12, 2023, please.

**Comment by anuani21** [comment]
@jjscheel, Based on the coverage matrix, I have started writing tests for native trigger behavior.I will able to share the draft by next meeting.

**Comment by jjscheel** [comment]
@anuani21, may I please get an update here?

**Comment by anuani21** [comment]
@timsifive ,Can you please guide me, how to run the tests in native trigger behaviour?

**Comment by anuani21** [comment]
@jjscheel ,I have generated the test for instruction count trigger(type 3) based on the  coverage matrix.

@timsifive, Can you please guide me, how to run the tests in native trigger behaviour in spike?

**Comment by pdonahue-ventana** [comment]
timsifive is now @rtwfroody

**Comment by anuani21** [comment]
@rtwfroody , Can you please guide me, how to run the tests in native trigger behaviour in spike?

**Comment by rtwfroody** [comment]
There's a README that describes how to run the tests: https://github.com/riscv-software-src/riscv-tests/blob/master/debug/README.md

If that's not clear, let me know.

**Comment by jjscheel** [comment]
@anuani21, please provide a status update when we have a moment.  As I mentioned in the meeting, this work will gate ratification of the specification if not complete in the next 2-3 months.

**Comment by anuani21** [comment]
@jjscheel, I will complete the work within  the end of April month.

Here are the updates from IITM,

Based on the coverage matrix I am generating the test cases for trigger type 2 (match control) and type 3(instruction count) with different set of arguments.

I will share the draft to @rtwfroody  for review shortly.

**Comment by pdonahue-ventana** [comment]
> Based on the coverage matrix I am generating the test cases for trigger type 2 (match control) and type 3(instruction count) with different set of arguments.

We should focus on type 6 rather than type 2.  Type 2 says: "This trigger type is deprecated. It is included for
backward compatibility with version 0.13."  Type 6 says: "This replaces mcontrol in newer implementations and serves to provide additional functionality."

**Comment by jjscheel** [comment]
Thanks, @anuani21 and @pdonahue-ventana!

**Comment by anuani21** [comment]
@paul, we need to write the test cases for trigger type 6 and trigger type
3 alone.

**Comment by anuani21** [comment]
@jjscheel, Based on the coverage matrix I am generating the test cases for trigger type 6 (match control6) and type 3(instruction count) with different set of arguments.

**Comment by anuani21** [comment]
@jjscheel, Based on the coverage matrix I have generated the test case for trigger type 6(match control 6),

-where a trigger is set and when it is hit the hart takes a debug exception with the following set of arguments
     -Execute=1 match =0, when trigger is hit,it perform the following action, value of tdata1 equals tdata2.
     -Execute=1 match =1, when trigger is hit,it perform the following action, top M-bits of tdata1 equals with the top m-bits of tdata2.(NAPOT)
     -Execute=1 match =2, when trigger is hit,it perform the following action, the value of tdata1 is greater than or equal to tdata2
     -Execute=1 match =3, when trigger is hit,it perform the following action, the value of tdata1 is lesser than or equal to tdata2.
     -Execute=1 match =4, when trigger is hit,it perform the following action, mask low operation.
     -Execute=1 match =5, when trigger is hit,it perform the following action, mask high operation.

I will share the above tests and logs to @rtwfroody for review shortly.

In trigger type 6, the following tests are pending 
- load=1 with different set of match values(0,1,2,3,4,5)
- store =1 with different set of match values(0,1,2,3,4,5)

I will complete the pending tests before the next meeting.

**Comment by jjscheel** [comment]
@anuani21, this is great progress.  Am I correct in understanding that work remains on trigger type = 3?

**Comment by rtwfroody** [comment]
> I will share the above tests and logs to @rtwfroody for review shortly.

I never saw this. Did I miss an email somewhere?

**Comment by anuani21** [comment]
Hi,

I had completed the test cases for store and execute with different match
values ranges from 0 to 5 in trigger type 6.

In load operation, I am facing some issues, once it complete I will share
the entire trigger type 6 test cases for review.

**Comment by anuani21** [comment]
@jjscheel,

Updates from IITM,

- Based on the coverage matrix, I have generated test cases for match control( type 6) triggers, where the trigger is set with parameters (store=1 and load=1 and set of match values[0,1,2,3,4,5]) when the trigger is hit , then the hart takes a debug exception and it perform different action based on the match values.

  - I had written separate tests cases for different match values.

The pending test cases are,

- Type3 trigger test cases.

- In Type6 trigger when load =1 and match=0  facing some issue in this test cases, I will fix this issue.

when the above issue fixed in type 6 trigger, i will share the entire trigger 6 test cases  and logs for review to @rtwfroody.

**Comment by anuani21** [comment]
@jjscheel,
Updates from IITM , 

- MControl6 ( type 6 ) triggers have been implemented but experiencing version related issues while implementing load operation in type 6, with regard to this reached out to @pdonahue-ventana  .

-Paul suggested checking tinfo.version bit upon  checking tinfo.version == 0 which means it is running in older version ( 0.13 ) , but the debug spec says to run in newer version ( 1.0 ) in which tinfo.version = 1 . 

-After updating both spike and openocd to the latest commit , raised the same issue in openocd (https://github.com/riscv-collab/riscv-openocd/issues/1053) to which got the response stating this could be a spike issue.

-Raised the same issue  in spike ( https://github.com/riscv-software-src/riscv-isa-sim/issues/1651#issuecomment-2084430295) . 
This issue seems not yet to be addressed in the spike development forum .

Can we get additional help regarding this issue ?

- Only after this issue has been resolved , we can  move forward and implement all triggers in the newer version.
- Have sent the test case and logs for trigger 6 (with  load store and execute operations ) in older version  for review to @rtwfroody.
- Type3 trigger test cases are pending and expected to complete before the next meeting.

**Comment by jjscheel** [comment]
@pdonahue-ventana, @rtwfroody, please see the past post and advice @anuani21 on how to work past the issue.

**Comment by rtwfroody** [comment]
I'll update the spike trigger implementation to implement hit0/hit1 instead of the timing bit for mcontrol6. Should be done by May 15.

> Have sent the test case and logs for trigger 6 (with load store and execute operations ) in older version for review to @rtwfroody.

How did you send these? I haven't seen them.

**Comment by anuani21** [comment]
Tims, can you please send your mail address so that i can send it once
again.

**Comment by rtwfroody** [comment]
My email address is tim@casualhacker.net.

**Comment by anuani21** [comment]
@jjscheel,

Updates from IITM , 

Following the review done by @rtwfroody ,by ongoing work in debug task 

1) We are waiting for the fix in spike for mcontrol6.
2) Parallelly we are going to revisit the written tests and follow a new flow.
Once the fix is in place we will resume to run the tests.

**Comment by rtwfroody** [comment]
I have a spike branch that updates the mcontrol6 to version 1.0. @anuani21, can you use this branch for your mcontrol6 tests? It should be correct, but since there are no tests there may be bugs. If it's behaving strangely, please send me the test you're working on and I'll look into it.

The pull request (which I won't pursue further until I get confirmation that this code passes the new tests) is at https://github.com/riscv-software-src/riscv-isa-sim/pull/1667.

**Comment by anuani21** [comment]
@rtwfroody, Fix for mcontrol6 which updates version to 1.0 is working fine.
@rtwfroody and @jjscheel, I have started revisit the written test for mconrol6 type trigger and other trigger types.

**Comment by YenHaoChen** [comment]
Hello there,

I provided the source code of trigger type 3~6 to Spike last year (based on type 2  originally provided by @rtwfroody).

Sorry that I overlooked the tinfo.version. Since you are using the trigger module in Spike. I want to say hi to you, and let you know that I am willing to provide additional commits for bugs in my code.

**Comment by rtwfroody** [comment]
@YenHaoChen's mcontrol6.hit{0,1} changes merged into spike mainline through https://github.com/riscv-software-src/riscv-isa-sim/pull/1257.

@anuani21, please use the latest version of spike for your testing. Since his implementation is different from mine, it might not act the same.

**Comment by jjscheel** [comment]
@anuani21, what is the latest status here?  When do you project having PRs to submit?

**Comment by anuani21** [comment]
Hi Jeff,

Due to some emergency situation, I am not available last week. From
yesterday I am started the work again.

I will make progress on this issue and update you.

**Comment by jjscheel** [comment]
No worries, @anuani21.  Hope all is well.  Thanks for the update.

**Comment by anuani21** [comment]
@jjscheel,

Updates from IITM , 

1)
Implemented mcontrol6 trigger with the latest spike version , raised a few issues with Tim . 
Tim has informed that the issues will be fixed by june 24th . We can move forward only after the fix.

Noted few more issue , hoping to receive some fix / clarification 

a) In the log generated , it is showing tcontrol could fetch register 'tcontrol' . 
While implementing trigger passing ebreak statement manually , mcause value 2 which is mapped to illegal instruction operation . 

b) In the log dcsr value is 0x4000b083 this results in the dcsr.ebreakm bit set to 1 , which is mapped to debug mode . dcsr.ebreakm bit should be 0 , only then it will raised breakpoint exception 

2)
Parallel till then will work on Code Size Reduction , will pick up from the last update.

**Comment by jjscheel** [comment]
Thanks, @anuani21, for the update.

A couple questions:

1. Once we have the Spike fixes, are there any other failures in the existing tests?
2. Have we written **all** tests we will need at this time?

**Comment by anuani21** [comment]
Only type 3 trigger tests are pending. I am writing the tests. For these
test also we need the spike fix.

Without spike fix, I am not getting expected result for mcause and mepc

**Comment by jjscheel** [comment]
What is your target date for completing the type 3 triggers tests?

**Comment by anuani21** [comment]
Within this week I will complete the writing type 3 trigger tests.

**Comment by jjscheel** [comment]
Ok, thank you.  I ask because we will need all tests written and coverpoints provided for review to be able to allow the Debug extension to proceed to Ratification.  I am trying to prevent this from being the last piece of work needed.  Please let us know when you think all test cases and coverpoints are done per the SOW (first entry in issue) definition.

Thanks for your attention.

**Comment by anuani21** [comment]
Ok i will make it ready

**Comment by anuani21** [comment]
@jjscheel 

All test  cases for type 3 icount trigger and type 6 mcontrol6 trigger with different configuration(execute , load , store and match =0,1,2,3,4,5) are ready , only facing few issues like expected result/values is not yet achieved ( in mcause , mtval and mepc register).

Fix for tcontrol is ready , but we are asked some more clarification with Tim on the same.

**Comment by gsterlin** [comment]
@anuani21 Allen is asking if you can provide the exact questions that you need answered, and he can help work with Tim to better understand what is being asked.

Can you also provide a pointer to the exact tests, and do you have a PR/repo that you can point us to so that we can ensure we are talking about the same content?

We expect that spike and sail should be providing the same answers.  Are one or the other missing this result which is causing the confusion, or is there a different issue with the test results?

Thanks!

**Comment by anuani21** [comment]
Previous Update:

1 As per Tim suggested,we are started and run the test in riscof framework.
2. Currently,we dont have test yet. we will make a pointer once the test is ready.


Curent update:
1.We have ported sdtrig extension in riscof framework and riscv-config.
2. Written a sample test and run it in the riscof framework, it works fine.
3.I have to write  all tests with trigger types 6 and type 3. it will complete by the end of the month.

**Comment by anuani21** [comment]
@jjscheel,

Current update:
1.I have the written the test cases for mcontrol6 trigger type for execute,load and store operations.
2.Having problems in read/write the CSR_TDATA1 looped with Tim.

Pending work;
1.Rewrite the trigger type3 test cases as per ACT framework.
2.Include a new fix  to generate the coverpoints in riscv-isac and merge them with coverage report.

**Comment by rtwfroody** [comment]
@anuani21 When you update status here, can you link to your in-progress work? It's really helpful as a reference when you have questions, and also to get more detail about the exact state of things.

**Comment by anuani21** [comment]
@rtwfroody,

Here I have attached the test in the link https://gitlab.com/ptprasanna/actreports/-/commit/d1b0826b0342545d2528495502c7807f5272d4c 

When I am running the test, writes some value in tdata1 register  and read back the value from tdata1 it shows only 0.Can you please review it an tell me the comments??

**Comment by rtwfroody** [comment]
That's not what I see. What I did:
1. Ran riscof with a custom test list to get the test compiled: `riscof run --config=config.ini --suite=riscv-arch-test/riscv-test-suite/ --env=riscv-arch-test/riscv-test-suite/env --testfile test_list.yaml`
2. The test hangs, so ^C
3. Dig out the test file, and run it manually with spike: `spike -d --isa=rv32imc +signature=/home/tnewsome/projects/riscv/riscof/riscof_work/rv32i_m/Sdtrig/src/mcontrol6_read.S/dut/DUT-spike.signature +signature-granularity=4 my.elf`
4. In the debugger, run until writing tdata1: `untiln pc 0 0x80000250` (I imagine that value might depend on compiler or whatever)
```
core   0: 0x8000024e (0x00000641) c.addi  a2, 16
(spike)    
core   0: 0x80000250 (0x7a161073) csrw    tdata1, a2
(spike) 
core   0: 0x80000254 (0x7a1026f3) csrr    a3, tdata1
(spike) reg 0 a2
0x60001010
(spike) reg 0 a3
0x60000010
```

That all looks OK. The difference in input/output is because CSR_MCONTROL6_M and CSR_MCONTROL6_EXECUTE are already shifted. There are several kinds of macros defines, and you're confusing 2 of them. (In my example I also confused them, but I got lucky because I think all the mistakes ended up being 0.)
CSR_<csr name>_<field name> is the mask of that field. It's already offset. E.g. CSR_MCONTROL6_M
CSR_<csr name>_<field name>_<field value> is just a field value. It's not offset. E.g. CSR_TDATA1_DMODE_BOTH

**Comment by rtwfroody** [comment]
@anuani21, that's great! With some of the tests complete, can you start a PR to submit those passing tests to riscv-arch-test? I suspect there will be some back and forth to get things to their standards (whatever they are), and I'd like to learn that process sooner rather than later.

**Comment by YenHaoChen** [comment]
A recent discussion talks about the existence of tcontrol.
https://github.com/riscv-software-src/riscv-isa-sim/pull/1742

**Comment by anuani21** [comment]
@jjscheel,

Updates from IITM,

1] Action breakpoint bug was resolved by Tim ( https://github.com/rtwfroody/riscv-isa-sim/tree/tcontrol_mte ), now getting trigger and mcause value 3
2] Getting incorrect value in HIT bit , Looped with Tim on the problem we are facing  (incorrect value being set in HIT bit)
3] So far all values are correct as expected other than hit bit. Link attached to  load test cases for  mcontrol6 
 https://gitlab.com/ptprasanna/actreports/-/tree/main/Native%20Trigger/LOAD?ref_type=heads
4] PR to be raised shortly

**Comment by jjscheel** [comment]
Sounds like we are getting close.  Keep up the good work!

**Comment by anuani21** [comment]
Updates from IITM,

- I have written  all test cases for icount trigger and mcontrol6 trigger for RV64 also.
- I am facing some issue in the following test,
       -In store and execute operations, some of the test cases while writing and reading the tdata1 and tdata2 register,the values are not getting correctly.
       - In load operation during data match trigger, both hit bit are setting up.Need to fix this issue.
 - During address match trigger in load operation, hit bit are set up correctly. For these test cases, I have raised a PR in riscv-arch-test.
          https://github.com/riscv-non-isa/riscv-arch-test/pull/487
 - Once issue fixed, I will add other test cases to this PR.

**Comment by anuani21** [comment]
@jjscheel,

1.After reviewing the PR, I had got a some comments from TIm and Allen.
2.Based on the comments,I have done some changes in the test and run the test without tcontrol because by default , s-mode is implemented.
3. After changes, I had run some of the tests and it is working fine.I will update it in the PR shortly.
4. I need  to run the updated tests in RV64.

**Comment by jjscheel** [comment]
Thank, you @anuani21.  It sounds like we are very close.  What's left besides running the tests in Rv64?

**Comment by rtwfroody** [comment]
> 2.Based on the comments,I have done some changes in the test and run the test without tcontrol because by default , s-mode is implemented.

Note that these tests need to run on all RISC-V implementations that implement Sdtrig. That includes ones that implement S-Mode, and ones that do not. We're just using spike to develop the tests because it's the easiest target to work with.

**Comment by anuani21** [comment]
@jjscheel,

I have updated the latest test in the PR.

**Comment by jjscheel** [comment]
Thanks, @anuani21.  How are we doing running tests on RV64?  Do our tests run in multiple modes, including S-mode as @rtwfroody has noted?

**Comment by anuani21** [comment]
In Rv64 it is working fine. Yes our test case will run with S-mode and
without S-mode.

**Comment by jjscheel** [comment]
@rtwfroody, @pdonahue-ventana, when Anusha completes creates her PR(s), we can clear the ACT Freeze Waiver by a review and confirmation that they appear feature complete.

**Comment by rtwfroody** [comment]
We don't need to wait for the PR to be merged?

**Comment by jjscheel** [comment]
No.  We need "feature complete" PR for Freeze.

**Comment by anuani21** [comment]
@jjscheel,

As per the Tim Comments, I had modified the test cases and updated it in the PR.

**Comment by jjscheel** [comment]
Or, @anuani21, a better question would be is this PR the only one? https://github.com/riscv-non-isa/riscv-arch-test/pull/487

**Comment by anuani21** [comment]
@jjscheel,

All tests are updated n the PR, Awaiting for a reply.

**Comment by jjscheel** [comment]
Thanks, @anuani21.  I talked to @pdonahue-ventana at Summit last week and he may be able to provide a review here as @rtwfroody is quite busy.  They understand that you are waiting on them.

**Comment by jjscheel** [comment]
@anuani21, do all tests pass the expected coverage?

**Comment by anuani21** [comment]
Yes Jeff ,Tests passed with the expected coverage.

**Comment by jjscheel** [comment]
@rtwfroody and @pdonahue-ventana, thanks for your reviews and comments on the PR.  If you've completed the review of all code in the PR, the first question we need to answer is whether we believe the test are functionally complete.  If so, we can conclude the Freeze waiver for ACT.  Can you share your thoughts/comments here?  If your answer is no, please clearly identify what you'd like to see resolved to be considered complete.

**Comment by pdonahue-ventana** [comment]
If you thought that RISC-V has lots of optional features, the debug spec has exponentially more options.  @rtwfroody and I recently discussed how far we want to go down the path of testing the seemingly endless combinations of possible things.  The two of us agreed that the ACT PR contains a solid baseline of testing and we're OK with moving forward with that.

**Comment by jjscheel** [comment]
@pdonahue-ventana and @rtwfroody, thanks for your feedback.  I will consider the Freeze Waiver work for ACT complete.

@anuani21, please continue working on review comments and trying to get the PR merged.  You're making GREAT progress!!!

**Comment by anuani21** [comment]
@jjscheel,

Icount test cases in the PR are still under review.

**Comment by anuani21** [comment]
@jjscheel,

Changes done in the test case as requested by Tim.Updated test cases in the PR. Waiting for approval.

**Comment by JAYANTH-IITM** [comment]
Waiting for review and ratification

**Comment by jjscheel** [comment]
Sdtrig work is now owned by the ACT SIG.  See the RISC-V [Certification Implementation Tracking](https://docs.google.com/spreadsheets/d/1EuG4LV1TQDx4wTzYDWEAvq0fwaWjNuxenwDARZmrN-E/edit?gid=0#gid=0) spreadsheet for details on it's completion.

HMC will be delivering the tests.

Closing.
